### PR TITLE
Hosts-file rollback hardening on partial failures

### LINKED ISSUE #310: Hosts-file rollback hardening on partial failures
## Summary
Harden hosts-file update rollback behavior when blocking writes fail partially.

## Scope
- Review blocker write pipeline for partial-write and interruption paths.
- Guarantee atomic-like recovery semantics using staged backups and explicit rollback.
- Add tests for failure injection across write/restore boundaries.

## Acceptance criteria
- Failed blocking updates always restore the previous hosts-file state.
- No orphaned focustime block sections remain after rollback scenarios.

### PR DESCRIPTION
**Pull request by utilForever**
## What

Harden the hosts backend write pipeline so partial failures during block/unblock updates trigger explicit rollback to the prior hosts-file state.

## Why

Issue #310 requires stronger recovery guarantees when hosts updates fail partway through writing or replacement, so focustime does not leave orphaned block sections behind.

Closes #310.

### Required

- [x] `cargo check --all` passes
- [x] `cargo fmt --all -- --check` passes
- [x] `cargo clippy --all-targets -- -D warnings` passes
- [x] `cargo test --all` passes
- [x] I linked the related issue (for example: `Closes #25`)

### Functional Validation

- [ ] Pomodoro timer behavior was verified for this change (if applicable)
- [x] Site blocking behavior was verified for this change (if applicable)
- [ ] WakaTime integration behavior was verified for this change (if applicable)
- [x] I added or updated tests for changed behavior (if applicable)

### Configuration & Docs

- [ ] User-facing docs were updated (`README.md` or relevant docs, if applicable)
- [ ] New dependencies/configuration are documented (if applicable)
- [x] No sensitive values or credentials were introduced

### If Applicable

- [ ] Security impact considered (run `cargo audit` locally if needed)
- [ ] Breaking behavior changes are clearly described in this PR

## Summary by CodeRabbit

* **Bug Fixes**
  * Hosts-file updates now use transactional write mechanisms with rollback capability on failure.

* **Tests**
  * Added test coverage for hosts-file update failure scenarios and rollback behavior.

[Review Change Stack](https://app.coderabbit.ai/change-stack/utilForever/focustime/pull/336?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/blocker.rs [hunk 1]
```diff
@@ -657,19 +673,27 @@ impl SiteBlocker {
...
-        let original = fs::read_to_string(HOSTS_FILE)?;
+        self.apply_hosts_block_to_path(Path::new(HOSTS_FILE))
+    }
+
+    fn apply_hosts_block_to_path(&self, hosts_path: &Path) -> io::Result<()> {
+        let original = fs::read_to_string(hosts_path)?;
         let content = self.build_blocked_hosts_content(&original);
-        atomic_write_hosts(&content)?;
+        atomic_write_hosts_to_path(hosts_path, &content)?;
         flush_dns_cache();
```
[the remark this answers is not in the document]
Comment (utilForever) [reply src/blocker.rs]: Implemented in commit bf704d4. I reordered the hosts mutation flow so block/unblock content is now derived from the transaction snapshot content, eliminating the pre-snapshot TOCTOU window.

File: src/blocker.rs [hunk 2]
```diff
@@ -951,31 +975,172 @@ fn probe_hosts_write_path(path: &Path) -> io::Result<()> {
...
+            return rollback_hosts_write(hosts_path, &snapshot_path, &staged_path, error);
+        }
+    }
+
+    if let Err(error) = maybe_fail_hosts_write_step(HostsWriteFailStep::AfterReplace) {
+        return rollback_hosts_write(hosts_path, &snapshot_path, &staged_path, error);
+    }
+
+    remove_file_if_exists(&staged_path)?;
+    remove_file_if_exists(&snapshot_path)?;
```
[the remark this answers is not in the document]
Comment (utilForever) [reply src/blocker.rs]: Implemented in commit fae8f1a. Post-commit cleanup for staged/snapshot temp files is now best-effort with warning logs, so successful replacements no longer return failure due only to temp-file cleanup errors.
